### PR TITLE
ENH: Test NIFTI I/O with small voxels

### PR DESCRIPTION
**Pull request by cookpa**
This covers a previous bug where the sform would not be read if the
determinant of the sform matrix was < 1E-5.

Added some more error messages to the itkNiftiReadWriteDirectionTest
so that it's clearer where any failures occur.

## PR Checklist
- [X] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [X] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [X] Added test (or behavior not changed)
- [X] Updated API documentation (or API not changed)
- [X] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [X] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [X] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKSphinxExamples) for all new major features (if any)

Refer to the [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) for
further development details if necessary.

### CODE REVIEW

**Comment by cookpa** [comment]
This tests the fix in #2701, many thanks to @jhlegarreta for guidance on writing the tests.

File: Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx [hunk 1]
```diff
@@ -61,14 +61,16 @@ itkNiftiReadWriteDirectionTest(int ac, char * av[])
...
     return EXIT_FAILURE;
   }
   // Verify that the qform alone is not identical to the sform (due to lossy storage capacity)
   // This difference is small, but periodically causes numerical precision errors,
   // VERIFY THAT THE DIRECTIONS ARE ACTUALLY DIFFERENT, They are expected to be different due to lossy
-  // representaiton of the qform.
+  // representation of the qform.
   if (LPSLabel_direction == LPSLabel_nosform_direction)
   {
+    std::cerr << "Error: direction matrices from sform and qform are identical" << std::endl;
```
Comment (jhlegarreta) [inline Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx:73]: In order to convey a more clear information about the error rationale, I would rephrase this as:
```
std::cerr << "Error: direction data must be different for qfrom- and sform-complete and qform-only versions of the same image due to lossy storage capacity." << std::endl;
```

@cookpa Do you think it is worthwhile adding some further information about the failure ? Something like, e.g.
```
std::cerr << "qfrom and sform-complete image direction: " << LPSLabel_direction << std::endl;
std::cerr << "qform-only image direction:" << LPSLabel_nosform_direction << std::endl;
```

File: Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx [hunk 2]
```diff
@@ -61,14 +61,16 @@ itkNiftiReadWriteDirectionTest(int ac, char * av[])
   // The sforms should be identical in both cases.
   if (LPSLabel_direction != LPSLabel_noqform_direction)
   {
+    std::cerr << "Error: sform direction matrices differ when qform is not present" << std::endl;
```
Comment (jhlegarreta) [inline Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx:64]: In order to convey a more clear information about the error rationale, I would rephrase this as:
```
std::cerr << "Error: direction data must be identical for qfrom- and sform-complete and sform-only versions of the same image." << std::endl;
```

@cookpa Do you think it is worthwhile adding some further information about the failure ? Something like, e.g.
```
std::cerr << "qfrom- and sform-complete image direction: " << LPSLabel_direction << std::endl;
std::cerr << "sform-only image direction:" << LPSLabel_noqform_direction << std::endl;
```

File: Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx [hunk 3]
```diff
@@ -107,6 +109,7 @@ itkNiftiReadWriteDirectionTest(int ac, char * av[])
   }
   if (!is_close_qform_converted)
   {
+    std::cerr << "Error: sform reconstructed from qform is not similar to qform" << std::endl;
     std::cerr << LPSLabel_nosform_direction << " \n\n" << reread_direction << std::endl;
```
Comment (jhlegarreta) [inline Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx:113]: Maybe this last print message could be improved to report what each variable corresponds to. Something like, e.g.
```
std::cerr << "Original qform-only image direction: " << LPSLabel_nosform_direction << std::endl;
std::cerr << "Original qform-only image re-written as qfrom and sform-complete image direction:" << reread_direction << std::endl;
```

**Review by jhlegarreta — COMMENTED** [review]
:100: Thanks for doing this @cookpa . Suggesting a few changes inline that IMO would improve the printed error reports.

**Comment by cookpa** [comment]
Thanks @jhlegarreta I made some more changes. I also changed the variable names to be more generic, instead of referencing LPSLabels.nii.gz.

**Review by jhlegarreta — APPROVED** [review]
Thanks for considering the changes @cookpa.

Pushed one commit to make the style of the test match the ITK conventions (you may want to have a look at appendix 3 in https://itk.org/ItkSoftwareGuide.pdf), and remove some duplicate/unnecessary aliases and variable, and another one to fix a typo.

File: Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx (hunk 2 above)
Comment (jhlegarreta) [reply Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx]: Addressed in 7f16366.

File: Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx (hunk 3 above)
Comment (jhlegarreta) [reply Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx]: Addressed in 7f16366.

File: Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx (hunk 1 above)
Comment (jhlegarreta) [reply Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx]: Addressed in 7f16366.

File: Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx [hunk 4]
```diff
@@ -49,28 +49,34 @@ itkNiftiReadWriteDirectionTest(int ac, char * av[])
   };
 
   using TestImageType = itk::Image<float, 3>;
-  TestImageType::Pointer LPSLabel = reader_lambda(av[1]);
-  TestImageType::Pointer LPSLabel_noqform = reader_lambda(av[2]);
-  TestImageType::Pointer LPSLabel_nosform = reader_lambda(av[3]);
+  TestImageType::Pointer InputImage = reader_lambda(av[1]);
```
Comment (dzenanz) [inline Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx]: It would be good if this commit was squashed with the previous one.

File: Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx [hunk 5]
```diff
@@ -40,85 +42,86 @@ itkNiftiReadWriteDirectionTest(int ac, char * av[])
   }
 
   using TestImageType = itk::Image<float, 3>;
-  auto reader_lambda = [](std::string filename) -> TestImageType::Pointer {
+  auto ReaderLambda = [](std::string filename) -> TestImageType::Pointer {
```
Comment (dzenanz) [inline Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx:45]: This lambda is no longer necessary (see below).

File: Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx [hunk 6]
```diff
@@ -40,85 +42,86 @@ itkNiftiReadWriteDirectionTest(int ac, char * av[])
...
+    reader->SetFileName(filename);
     reader->Update();
     return reader->GetOutput();
   };
 
-  using TestImageType = itk::Image<float, 3>;
-  TestImageType::Pointer InputImage = reader_lambda(av[1]);
-  TestImageType::Pointer InputImage_noqform = reader_lambda(av[2]);
-  TestImageType::Pointer InputImage_nosform = reader_lambda(av[3]);
+  TestImageType::Pointer inputImage = ReaderLambda(av[1]);
```
Comment (dzenanz) [inline Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx:52]: why not use `TestImageType::Pointer inputImage = itk::ReadImage<TestImageType>(av[1]);`? ReadImage is now part of C++ ITK.

File: Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx [hunk 7]
```diff
@@ -40,85 +42,86 @@ itkNiftiReadWriteDirectionTest(int ac, char * av[])
...
   // Write image that originally had no sform direction representation into a file with both sform and qform
-  const std::string                            test_output_dir = av[4];
-  const std::string                            test_filename = test_output_dir + "/test_filled_sform.nii.gz";
+  const std::string                            testOutputDir = av[4];
+  const std::string                            testFilename = testOutputDir + "/test_filled_sform.nii.gz";
   itk::ImageFileWriter<TestImageType>::Pointer writer = itk::ImageFileWriter<TestImageType>::New();
-  writer->SetFileName(test_filename);
-  writer->SetInput(InputImage_nosform);
-  writer->Update();
+  writer->SetFileName(testFilename);
```
Comment (dzenanz) [inline Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx:87]: why not `ITK_TRY_EXPECT_NO_EXCEPTION(itk::WriteImage(inputImageNoSform, testFilename));`?

**Review by dzenanz — COMMENTED** [review]
I would prefer a single commit with Jon as co-author. But it is totally fine to have 2 commits with separate authors.

**Comment by jhlegarreta** [comment]
@cookpa let us know if you want to have a go at what @dzenanz suggests. Thanks.

**Comment by cookpa** [comment]
I think I can do this. If I understand correctly,

1. Test and push the changes proposed above
2. Squash commits on this PR to a single one, with you and I as authors in the trailer

I'll have a go later

**Comment by jhlegarreta** [comment]
> I think I can do this. If I understand correctly,
> 
>     1. Test and push the changes proposed above
> 
>     2. Squash commits on this PR to a single one, with you and I as authors in the trailer
> 
> 
> I'll have a go later

Yes. `git` will manage the authorship, but I am fine if somehow only yours is kept. My contribution is not the essential part in all this.

File: Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx (hunk 4 above)
Comment (cookpa) [reply Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx]: I've squashed the commits into one

File: Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx (hunk 5 above)
Comment (cookpa) [reply Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx]: Thanks for this, I replaced it

File: Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx (hunk 6 above)
Comment (cookpa) [reply Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx]: Using ReadImage throughout now

File: Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx (hunk 7 above)
Comment (cookpa) [reply Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx]: Done, thanks

File: Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx (hunk 4 above)
Comment (jhlegarreta) [reply Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx]: :+1: Done in ceac451.

File: Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx (hunk 5 above)
Comment (jhlegarreta) [reply Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx]: :+1: Done in ceac451.

File: Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx (hunk 6 above)
Comment (jhlegarreta) [reply Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx]: :+1: Done in ceac451.

File: Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx (hunk 7 above)
Comment (jhlegarreta) [reply Modules/IO/NIFTI/test/itkNiftiReadWriteDirectionTest.cxx]: :+1: Done in ceac451.

**Review by jhlegarreta — APPROVED** [review]
:100: thanks @cookpa.
